### PR TITLE
Add XML namespace support for oauth2Login()

### PR DESCRIPTION
**Pull request by jgrandja**
Fixes gh-4557

### CODE REVIEW

**Comment by jgrandja** [comment]
@rh-id Here is the initial draft ready for you to take over.

Take a look at the 2nd commit and you will see what needs to be implemented.

There is one [test](https://github.com/spring-projects/spring-security/pull/7743/files#diff-355c215b3327f2d76f1a9268b69505b2R44) that fails but should pass after the `BeanDefinitionParser`'s are implemented.

To start things off, I would suggest that you get familiar with `HttpSecurity.oauth2Login()` first since this is the equivalent implementation required for the xml namespace support. If you know this well than the xml namespace implementation should make things easier for you.

Here are a couple of tests to check out:

`OAuth2LoginApplicationTests`  - spring-security-samples-boot-oauth2login

`org.springframework.security.config.annotation.web.configurers.oauth2.client.OAuth2LoginConfigurerTests`


Let me know if you have any questions.

File: config/src/test/resources/org/springframework/security/config/http/OAuth2LoginBeanDefinitionParserTests-SingleClientRegistration.xml [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+<b:beans xmlns:b="http://www.springframework.org/schema/beans"
+		xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+		xmlns="http://www.springframework.org/schema/security"
+		xsi:schemaLocation="
+			http://www.springframework.org/schema/security
+			https://www.springframework.org/schema/security/spring-security.xsd
+			http://www.springframework.org/schema/beans
+			https://www.springframework.org/schema/beans/spring-beans.xsd">
+
+	<http auto-config="true">
```
Comment (jgrandja) [inline config/src/test/resources/org/springframework/security/config/http/OAuth2LoginBeanDefinitionParserTests-SingleClientRegistration.xml]: @rwinch @jzheaux When you have a moment, please review this `oauth2Login()` xml sample configuration and provide feedback. This sample config explicitly configures all possible configurations. However, most are optional as defined in the schema. Also, `issuer-uri` is also defined in the schema but it's not in this xml instance.

**Comment by rh-id** [comment]
hi @jgrandja , I had push to https://github.com/jgrandja/spring-security/pull/2 should i just push to this project instead?

There are some questions that I might need to ask.
1. the class "SecurityFilters" and "FilterComparator". "SecurityFilters" enum stored the filter and order information while "FilterComparator" had filter ordering as comparator. does it mean that we have to maintain both classes so that it will have the same filter order?

2. I assume both "BeanDefinitionParser" classes will need to be decoupled from "org.springframework.security.oauth2.client" package hence I will need to use class name instead of class instance itself as parameter when instantiating "RootBeanDefinition" right? if so then ClientRegistration might have problem since it doesn't have public constructor.

**Comment by jgrandja** [comment]
@rh-id The plan is to work off of this PR. You will see 2 commits in this PR with the initial work. In the 2nd commit "Add template for BeanDefinitionParser implementations", you will see instructions on what to implement. All updates should build off of these 2 commits. I'll close the PR you submitted to my fork.

> the class "SecurityFilters" and "FilterComparator". "SecurityFilters" enum stored the filter and order information while "FilterComparator" had filter ordering as comparator. does it mean that we have to maintain both classes so that it will have the same filter order?

`FilterComparator` is used for Java Config and it has already been updated as part of `HttpSecurity.oauth2Login()` so you shouldn't have to make any changes here. You will only need to update `SecurityFilters` since this is used for XML Config.

> I assume both "BeanDefinitionParser" classes will need to be decoupled from "org.springframework.security.oauth2.client" package hence I will need to use class name instead of class instance itself as parameter when instantiating "RootBeanDefinition" right? if so then ClientRegistration might have problem since it doesn't have public constructor.

I'm not sure I follow you here. It's not clear to me what the issue is. FYI, `ClientRegistration` can be instantiated using `withRegistrationId()` factory method.

**Comment by rh-id** [comment]
hi @jgrandja sorry I'm not sure if I get you as well. yes, I understand that we will work based on this PR and I had proceed to finish the implementation of the XML support and test it. 

should I create a PR to your fork ? or create  PR in this project? my latest commit at: https://github.com/rh-id/spring-security

and thanks for the info.

**Comment by jgrandja** [comment]
@rh-id 

> sorry I'm not sure if I get you as well

Ok lets simplify things. Please run the following:

`git fetch upstream pull/7743/head:oauth2login-xmlconfig`

NOTE: `upstream` in my setup is `git@github.com:spring-projects/spring-security.git`. It might be `origin` in your setup.

You should now have a local branch called `oauth2login-xmlconfig` that checks out this PR. You will notice 2 commits. Your work/commits should add on top of these 2 commits. After you have completed the work then you can simply submit a new PR. I'll close off this Draft PR.

Does this clarify things?

One important thing to mention is that [5.3.0.RC1](https://github.com/spring-projects/spring-security/milestones) is due Feb 5 and this feature needs to be merged by then. This gives approx. 3 1/2 weeks but I would like to see a complete implementation within the next 2 weeks if possible to provide enough time for review. Do you feel you have the time to complete based on these timelines? I don't want you to feel rushed or cause unnecessary stress if you have other job priorities. If you feel you cannot deliver with these timelines then that's fine as well and I'll take it over. Please let me know either way so we can co-ordinate.

**Comment by rh-id** [comment]
hi @jgrandja i had create the PR https://github.com/spring-projects/spring-security/pull/7821

**Comment by jgrandja** [comment]
Closing in favour of #7821

**Pull request closed without merging**